crates/sparrow-execution/src/source_tasks.rs

```diff
@@ -0,0 +1,65 @@
+use error_stack::ResultExt;
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use sparrow_arrow::Batch;
+use sparrow_scheduler::{Injector, PipelineInputs};
+
+use crate::Error;
+
+/// Manages the async streams that must be read for input.
+#[derive(Default)]
+pub(super) struct SourceTasks {
+    /// Streams to be read. Each stream should produce a `(batch, index)` pair,
+    /// with indices corresponding to the input in `read_inputs` to send the batch
+    /// to.
+    ///
+    // TODO: We may want to encapsulate this differently / better. Could we use
+    // `futures::sink(...)` to do this in some way? We may also want to allow
+    // the polling of these sources to be weighted towards the inputs that need
+    // a new batch to progress (eg., back-pressure).
+    reads: Vec<BoxStream<'static, (Batch, usize)>>,
+    /// For each read operation, the `PipelineInputs` to send the batches to.
+    read_inputs: Vec<PipelineInputs>,
+}
+
+impl SourceTasks {
+    pub fn add_read(
+        &mut self,
+        stream: impl futures::Stream<Item = Batch> + Send + Sync + 'static,
+        sink: PipelineInputs,
+    ) {
+        let index = self.reads.len();
+        self.read_inputs.push(sink);
+        self.reads
+            .push(stream.map(move |batch| (batch, index)).boxed());
+    }
+
+    /// Execute all of the sources.
+    ///
+    /// This currently joins them and runs them directly. Depending on the amount
```

Comment:
I suspect we'll need to change this. I believe that some sources (eg., Parquet) are potentially CPU intensive, so we probably want each source/stream to be run as a separate tokio task.